### PR TITLE
fetch exchange rate api should pick a working default

### LINKED ISSUE #551: fetch exchange rate api should pick a working default
on first startup , binance is the chosen default for the fetch exchange rate api, and if it's not available after three tries, then no exchange rate is selected.

It seems to be a better option for the fetch exchange rate function to try each available exchange api in an attempt to get a working api, and only when it's not able to find a working, it then defaults to none.

```
2024-06-24 08:45:26.414 [ERR] EXT: fetching ticker 1 failed: binance failed to fetch ticker for BTC-USDT: Get "https://api.binance.com/api/v3/ticker/24hr?symbol=BTCUSDT": dial tcp: lookup api.binance.com on 127.0.0.*: no such host. Retrying in 1s

2024-06-24 08:45:27.423 [ERR] EXT: fetching ticker 2 failed: binance failed to fetch ticker for BTC-USDT: Get "https://api.binance.com/api/v3/ticker/24hr?symbol=BTCUSDT": dial tcp: lookup api.binance.com on 127.0.0.*: no such host. Retrying in 2s

2024-06-24 08:45:29.437 [ERR] EXT: fetching ticker 3 failed: binance failed to fetch ticker for BTC-USDT: Get "https://api.binance.com/api/v3/ticker/24hr?symbol=BTCUSDT": dial tcp: lookup api.binance.com on 127.0.0.*: no such host. Retrying in 4s

2024-06-24 08:45:33.446 [ERR] EXT: binance: Error fetching ticker: binance failed to fetch ticker for BTC-USDT: Get "https://api.binance.com/api/v3/ticker/24hr?symbol=BTCUSDT": dial tcp: lookup api.binance.com on 127.0.0.*: no such host
```

### PR DESCRIPTION
**Pull request by vibros68**
Closes #551 

* Fetch exchange rate option just fetchs the choosen exchange by user and up to 3 times for fetching rate. If all times is failed. So `none` option is choosen
* This mr add change by fetch all other available exchange which is supported
* This mr also remove support for bittrex since bittrex is no longer serving

### CODE REVIEW

**Comment by dreacot** [comment]
```
2024-07-06 10:24:47.341 [ERR] EXT: fetching ticker 3 failed: binance failed to fetch ticker for LTC-USDT: Get "https://api.binance.com/api/v3/ticker/24hr?symbol=LTCUSDT": dial tcp: lookup api.binance.com: no such host. Retrying in 4s
2024-07-06 10:24:51.342 [INF] EXT: fetching from other exchanges
2024-07-06 10:24:52.498 [INF] EXT: messari is choosen
2024-07-06 10:24:52.500 [ERR] EXT: fetching ticker 1 failed: messari failed to fetch ticker for BTC-USDT: Get "https://api.binance.com/api/v3/ticker/24hr?symbol=BTCUSDT": dial tcp: lookup api.binance.com: no such host. Retrying in 1s
2024-07-06 10:24:53.502 [ERR] EXT: fetching ticker 2 failed: messari failed to fetch ticker for BTC-USDT: Get "https://api.binance.com/api/v3/ticker/24hr?symbol=BTCUSDT": dial tcp: lookup api.binance.com: no such host. Retrying in 2s
2024-07-06 10:24:55.507 [ERR] EXT: fetching ticker 3 failed: messari failed to fetch ticker for BTC-USDT: Get "https://api.binance.com/api/v3/ticker/24hr?symbol=BTCUSDT": dial tcp: lookup api.binance.com: no such host. Retrying in 4s
2024-07-06 10:24:59.508 [INF] EXT: fetching from other exchanges
2024-07-06 10:25:00.324 [INF] EXT: coinpaprika is choosen
2024-07-06 10:25:00.326 [ERR] EXT: fetching ticker 1 failed: coinpaprika failed to fetch ticker for DCR-USDT: Get "https://api.binance.com/api/v3/ticker/24hr?symbol=DCRUSDT": dial tcp: lookup api.binance.com: no such host. Retrying in 1s
2024-07-06 10:25:01.330 [ERR] EXT: fetching ticker 2 failed: coinpaprika failed to fetch ticker for DCR-USDT: Get "https://api.binance.com/api/v3/ticker/24hr?symbol=DCRUSDT": dial tcp: lookup api.binance.com: no such host. Retrying in 2s
2024-07-06 10:25:03.334 [ERR] EXT: fetching ticker 3 failed: coinpaprika failed to fetch ticker for DCR-USDT: Get "https://api.binance.com/api/v3/ticker/24hr?symbol=DCRUSDT": dial tcp: lookup api.binance.com: no such host. Retrying in 4s
2024-07-06 10:25:07.335 [INF] EXT: fetching from other exchanges
2024-07-06 10:25:07.619 [INF] EXT: messari is choosen
```

even though it says messari is chosen initially, it seems more api calls are still being made to try and determine a suitable default, and even after choosing messari, the UI still shows binance

Screenshot 2024-07-06 at 10 27 12 AM

and if one fails, i don't think there is any need to try it 3 times anymore, just move on to checking the next available api

**Comment by vibros68** [comment]
Okay

File: libwallet/ext/rate_source.go [hunk 1]
```diff
@@ -356,23 +351,38 @@ func (cs *CommonRateSource) fetchRate(market values.Market) *Ticker {
...
+		getTickerFn := cs.sourceGetTickerFunc(source)
 		select {
 		case <-cs.ctx.Done():
 			log.Errorf("fetching ticker canceled: %v", cs.ctx.Err())
 			return nil, cs.ctx.Err()
 		default:
-			newTicker, err = cs.getTicker(market)
+			newTicker, err = getTickerFn(market)
 			if err == nil {
+				log.Infof("%s is choosen", source)
```
Comment (dreacot) [inline libwallet/ext/rate_source.go:382]:
```suggestion
				log.Infof("%s is chosen", source)
```

**Review by dreacot — APPROVED** [review]

**Comment by dreacot** [comment]
please  with master, master contains dcrd v2.0 which has an upgraded database, so most devs are unable to test this PR

File: libwallet/ext/rate_source.go [hunk 2]
```diff
@@ -356,23 +351,38 @@ func (cs *CommonRateSource) fetchRate(market values.Market) *Ticker {
 }
 
 func (cs *CommonRateSource) retryGetTicker(market values.Market) (*Ticker, error) {
+	log.Infof("fetching %s rate from %v", market, cs.source)
```
Comment (dmigwi) [inline libwallet/ext/rate_source.go:354]: this line would be best suited to run just before: 
```
newTicker, err = cs.getTicker(market)
```

At line 362

File: libwallet/ext/rate_source.go [hunk 3]
```diff
@@ -356,23 +351,38 @@ func (cs *CommonRateSource) fetchRate(market values.Market) *Ticker {
...
+			return newTicker, nil
+		}
+	}
+	// fetch ticker from available exchanges
+	log.Infof("fetching from other exchanges")
+	for _, source := range sources {
+		if source == cs.source {
+			continue
+		}
+		log.Infof("fetching %s rate from %v", market, source)
```
Comment (dmigwi) [inline libwallet/ext/rate_source.go:373]: Same to this, move it to where the fetch is actually happening.

**Review by dmigwi — CHANGES_REQUESTED** [review]

**Comment by vibros68** [comment]
Done @dmigwi

**Review by dmigwi — APPROVED** [review]
LGTM

**Comment by dreacot** [comment]
@vibros68 please rebase with master https://github.com/crypto-power/cryptopower/pull/564#issuecomment-2213709051

File: libwallet/ext/rate_source.go [hunk 4]
```diff
@@ -141,7 +136,8 @@ type CommonRateSource struct {
 	tickers       map[values.Market]*Ticker
 	refreshing    bool
 	cond          *sync.Cond
-	getTicker     func(market values.Market) (*Ticker, error)
+	getTicker     tickerFunc
+	otherTickers  []tickerFunc
```
Comment (JustinBeBoy) [inline libwallet/ext/rate_source.go:140]: Remove otherTickers, I don't see anywhere use it

**Review by JustinBeBoy — CHANGES_REQUESTED** [review]

**Comment by JustinBeBoy** [comment]
@vibros68 please clean code and we can merge

**Comment by vibros68** [comment]
@JustinBeBoy done

**Review by JustinBeBoy — APPROVED** [review]

**Pull request merged**